### ISSUE TITLE
Windows x64 no-std build broken with errors related to compiler-builtins

### ISSUE THREAD

**Issue by byeongkeunahn** [rust-lang/rust#118559]
Hello, the nightly update on 2023-12-02 seems to break the continuous integration (CI) on the [basm-rs](https://github.com/byeongkeunahn/basm-rs) project on Windows x64 builds (nightly-x86_64-pc-windows-msvc). The Linux x86/x64 builds and the wasm32 builds continue to run fine. The recent commits on the repo is Linux-only and not related to Windows; thus the Windows build should not be affected. It seems like the recently merged pull request [113923](https://github.com/rust-lang/rust/pull/113923) may be relevant.

I expected to see this happen: the basm-rs CI build should continue to pass.

Instead, this happened: the basm-rs CI build fails with a linker error related to compiler-builtins.

### Meta
I tried to localize the error with `cargo-bisect-rustc` on Windows 11 x64 with the msvc toolchain as follows:
```
git clone https://github.com/byeongkeunahn/basm-rs.git
cargo-bisect-rustc -vv --start=2023-12-01 --end=2023-12-03
```
which results in the following analysis report:

searched nightlies: from nightly-2023-12-01 to nightly-2023-12-03
regressed nightly: nightly-2023-12-02
searched commit range: https://github.com/rust-lang/rust/compare/87e1447aadaa2899ff6ccabe1fa669eb50fb60a1...8c2b57721728233e074db69d93517614de338055
regressed commit: https://github.com/rust-lang/rust/commit/8c2b57721728233e074db69d93517614de338055

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>

Host triple: x86_64-pc-windows-msvc
Reproduce with:
```bash
cargo bisect-rustc --start=2023-12-01 --end=2023-12-03
```
</details>

The actual error message is about unresolved external symbols.
```
  = note: symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem4bcmp17hdd43ed361972df2aE
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6memcmp17h4000bdfb54398fd7E
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6memcpy17h76ab47a482abb17aE
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6memset17hddee021ef3511addE
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem6strlen17h8221a76238dd80f9E
          symbols.o : error LNK2001: unresolved external symbol _ZN17compiler_builtins3mem7memmove17h9a354b96e96d4c1eE
          ..(omitted)\target-bisector-ci-8c2b57721728233e074db69d93517614de338055-x86_64-pc-windows-msvc\x86_64-pc-windows-msvc\debug\deps\basm.exe : fatal error LNK1120: 6 unresolved externals
```

Thanks in advance!
